Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,480 @@
+{
+    auto lowercase_type = alternative.body_structure.type.to_lowercase();
+    auto lowercase_subtype = alternative.body_structure.subtype.to_lowercase();
+    return lowercase_type == "text" && lowercase_subtype.is_one_of("plain", "html");
+}
+
+void MailWidget::selected_mailbox()
+{
+    auto const& index = m_mailbox_list->selection().first();
+
+    if (!index.is_valid())
+        return;
+
+    auto& base_node = *static_cast<BaseNode*>(index.internal_data());
+
+    if (is<AccountNode>(base_node)) {
+        // FIXME: Do something when clicking on an account node.
+        return;
+    }
+
+    auto& mailbox_node = verify_cast<MailboxNode>(base_node);
+    auto& mailbox = mailbox_node.mailbox();
+
+    VERIFY(!(mailbox.flags & (unsigned)IMAP::MailboxFlag::NoSelect));
+
+    auto response = m_imap_client->select(mailbox.name)->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to select mailbox. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to select mailbox. The server says: '{}'", response.response_text()));
+        return;
+    }
+
+    auto fetch_command = IMAP::FetchCommand {
+        // Mail will always be numbered from 1 up to the number of mail items that exist, which is specified in the select response with "EXISTS".
+        .sequence_set = { { 1, (int)response.data().exists() } },
+        .data_items = {
+            IMAP::FetchCommand::DataItem {
+                .type = IMAP::FetchCommand::DataItemType::BodySection,
+                .section = IMAP::FetchCommand::DataItem::Section {
+                    .type = IMAP::FetchCommand::DataItem::SectionType::HeaderFields,
+                    .headers = { { "Subject", "From" } },
+                },
+            },
+        },
+    };
+
+    auto fetch_response = m_imap_client->fetch(fetch_command, false)->await().release_value();
+
+    if (response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to retrieve subject/from for e-mails. The server says: '{}'", response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve e-mails. The server says: '{}'", response.response_text()));
+        return;
+    }
+
+    Vector<InboxEntry> active_inbox_entries;
+
+    for (auto& fetch_data : fetch_response.data().fetch_data()) {
+        auto& response_data = fetch_data.get<IMAP::FetchResponseData>();
+        auto& body_data = response_data.body_data();
+
+        auto data_item_has_header = [](IMAP::FetchCommand::DataItem const& data_item, String const& search_header) {
+            return data_item.section.has_value()
+                && data_item.section->type == IMAP::FetchCommand::DataItem::SectionType::HeaderFields
+                && data_item.section->headers.has_value()
+                && data_item.section->headers->find_if([&search_header](String const& header) {
```

Comment:
If `headers` is a `Vector` (which I think it is), using `contains_slow` should be just fine. `find_if` won't be faster just because it doesn't tell you it's slow. :^)

Reply:
Also, please use multiple lines to make this code a bit more readable, so something like this:
```cpp
auto data_item_has_header = [](IMAP::FetchCommand::DataItem const& data_item, String const& search_header) {
    if (!return_data_item.section.has_value())
        return false;
    if (data_item_section->type != IMAP::FetchCommand::DataItem::SectionType::HeaderFields)
        return false;
    if (!data_item_section->headers.has_value())
       return false;
    return data_item_section->headers->contains_slow(search_header);
};
```